slack_bolt/workflows/step/async_step.py

```diff
@@ -9,11 +13,261 @@
 )
 from slack_bolt.middleware.async_custom_middleware import AsyncCustomMiddleware
 from slack_bolt.response import BoltResponse
-from slack_sdk.web.async_client import AsyncWebClient
+from .utilities.async_complete import AsyncComplete
 from .utilities.async_configure import AsyncConfigure
 from .utilities.async_fail import AsyncFail
-from .utilities.async_complete import AsyncComplete
 from .utilities.async_update import AsyncUpdate
+from ...error import BoltError
+from ...listener_matcher.async_listener_matcher import (
+    AsyncListenerMatcher,
+    AsyncCustomListenerMatcher,
+)
+from ...middleware.async_middleware import AsyncMiddleware
+
+
+class AsyncWorkflowStepBuilder:
+    callback_id: Union[str, Pattern]
+    _edit: Optional[AsyncListener]
+    _save: Optional[AsyncListener]
+    _execute: Optional[AsyncListener]
+
+    def __init__(
+        self, callback_id: Union[str, Pattern], app_name: Optional[str] = None,
+    ):
+        """This builder is supposed to be used as decorator.
+
+        my_step = AsyncWorkflowStep.builder("my_step")
+
+        @my_step.edit
+        async def edit_my_step(ack, configure):
+            pass
+
+        @my_step.save
+        async def save_my_step(ack, step, update):
+            pass
+
+        @my_step.execute
+        async def execute_my_step(step, complete, fail):
+            pass
+
+        app.step(my_step)
+
+        :param callback_id: the callback_id for the workflow
+        :param app_name: the application name mainly for logging
+        """
+        self.callback_id = callback_id
+        self.app_name = app_name or __name__
+        self._edit = None
+        self._save = None
+        self._execute = None
+
+    def edit(
+        self,
+        *args,
+        matchers: Optional[
+            Union[Callable[..., Awaitable[bool]], AsyncListenerMatcher]
+        ] = None,
+        middleware: Optional[Union[Callable, AsyncMiddleware]] = None,
+        lazy: Optional[List[Callable[..., Awaitable[None]]]] = None,
+    ):
+        """Register a new edit listener with details.
+        You can use this method as decorator as well.
+
+        @my_step.edit
+        def edit_my_step(ack, configure):
+            pass
+
+        It's also possible to add additional listener matchers and/or middleware
+
+        @my_step.edit(matchers=[is_valid], middleware=[update_context])
+        def edit_my_step(ack, configure):
+            pass
+        """
+        if len(args) == 1:
```

Comment:
This is the pattern for `@my_step.edit` (= the `edit` method is used without `()`).